src/Symfony/Component/Routing/Matcher/UrlMatcher.php

```diff
@@ -141,23 +153,31 @@ protected function matchCollection($pathinfo, RouteCollection $routes)
                 continue;
             }
 
-            // check HTTP method requirement
+            $hasRequiredScheme = !$route->getSchemes() || $route->hasScheme($this->context->getScheme());
             if ($requiredMethods = $route->getMethods()) {
                 // HEAD and GET are equivalent as per RFC
                 if ('HEAD' === $method = $this->context->getMethod()) {
                     $method = 'GET';
                 }
 
                 if (!in_array($method, $requiredMethods)) {
-                    if (self::REQUIREMENT_MATCH === $status[0]) {
+                    if ($hasRequiredScheme) {
                         $this->allow = array_merge($this->allow, $requiredMethods);
                     }
 
                     continue;
                 }
             }
 
-            return $this->getAttributes($route, $name, array_replace($matches, $hostMatches, isset($status[1]) ? $status[1] : array()));
+            $ret = $this->getAttributes($route, $name, array_replace($matches, $hostMatches, isset($status[1]) ? $status[1] : array()));
```

Comment:
what's the reason for doing that before the $hasRequiredScheme check?

Reply:
reverted, was a leftover from previous iteration